### ISSUE TITLE
False positive in single-line class

### ISSUE THREAD

**Issue by grodowski** [grodowski/undercover#216]
```
module Module
  class Bar < StandardError; end
end
```

Produces a warning on `Bar`, even though it's loaded and reported in coverage. Probably has to do with the `empty_def?` logic.

**Comment by splattael** [comment]
:wave: @grodowski  from https://gitlab.com/gitlab-org/quality/triage-ops/-/merge_requests/3113#note_2264745075

In https://gitlab.com/gitlab-org/quality/triage-ops/-/jobs/8674431352#L69 we are seeing a single line (not class definition) not being flagged as "covered".

Do you think this might be related to this issue? :shrug: 

```
🚨 WARNING: Coverage data is older than your latest changes and results might be incomplete. Re-run tests to update
undercover: 👮‍♂️ some methods have no test coverage! Please add specs for methods listed below
🚨 1) node `block (f)` type: block,
      loc: triage/triage/handler.rb:5:5, coverage: 0.0%
5: Dir[File.expand_path("../processor/**/*.rb", __dir__)].each { |f| require f } hits: n/a
Undercover finished in 0.1831s
```

`undercover` passes with `0.5.0` for this very same merge request.

**Comment by splattael** [comment]
@grodowski I only spotted now your comment that's related to all single-line code :sweat_smile: 

Anyway, I want to share the reproduction and the "git bisect" I did in https://gitlab.com/gitlab-org/quality/triage-ops/-/issues/1619#note_2265433256:

I can reproduce it on test project and tracked down the commit [036dd0ac6ad1f39cef5217573d5d261c1d2cf8bf](https://github.com/grodowski/undercover/commit/036dd0ac6ad1f39cef5217573d5d261c1d2cf8bf) via `git bisect`.

1. `git clone <undercover>`
2. `cd undercover`
3. `mkdir testing && cd testing`

4. `cat Gemfile`

```ruby
source "https://rubygems.org"

gem "rspec"
gem "simplecov"
gem "simplecov-lcov"
gem "undercover", path: "../"

5. `cat spec/spec_helper.rb`

```ruby
# the very top of spec_helper.rb

require 'simplecov'
require 'simplecov-lcov'

$LOAD_PATH << "."

SimpleCov::Formatter::LcovFormatter.config.report_with_single_file = true
SimpleCov.formatter = SimpleCov::Formatter::LcovFormatter
SimpleCov.start do
  add_filter(/^\/spec\//) # For RSpec
  add_filter(/^\/test\//) # For Minitest
end

require 'undercover'
```

6. `cat spec/foo_spec.rb`

```ruby
require "spec_helper"

require_relative "../test.rb"

RSpec.describe "foo" do
  it "works" do
    expect(true).to eq(true)
  end
end
```

7. `cat test.rb`

```ruby
Dir[File.expand_path("./foo/*.rb", __dir__)].each { |f| require f }
```

8. `git init && git add -A && git commit -m "Initial commit"`
9. Make a change in `test.rb`


```diff
diff --git a/test.rb b/test.rb
index 64beb8e..167c079 100644
--- a/test.rb
+++ b/test.rb
@@ -1 +1,3 @@
 Dir[File.expand_path("./foo/*.rb", __dir__)].each { |f| require f }
+
+# A change
```

10. `cd ../` and create a `script.sh` (`chmod a+x script.sh`) for `git bisect`:

```shell
#!/bin/bash

set -euo pipefail

mise use ruby@3.3.3

bundle
(cd testing && bundle && bundle exec undercover)

git reset --hard
```

11. Run bisect

```shell
git bisect start
git bisect good v0.5.0
git bisect bad v0.6.0
git bisect run ./script.sh

git bisect run ./script.sh
status: waiting for both good and bad commits
status: waiting for bad commit, 1 good commit known
Bisecting: 7 revisions left to test after this (roughly 3 steps)
[31fbda6f5393d7f270bdbfdce17279cb027d2782] Handle nil merge_base in Changeset
running  './script.sh'
mise ~/devel/gitlab/triage-ops/undercover/.tool-versions tools: ruby@3.3.3
Bundle complete! 11 Gemfile dependencies, 35 gems now installed.
Gems in the group 'production' were not installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
Bundle complete! 4 Gemfile dependencies, 20 gems now installed.
Gems in the group 'production' were not installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
🚨 WARNING: Coverage data is older than your latest changes and results might be incomplete. Re-run tests to update
undercover: ✅ No coverage is missing in latest changes
Undercover finished in 0.0039s
HEAD is now at 31fbda6 Handle nil merge_base in Changeset
Bisecting: 3 revisions left to test after this (roughly 2 steps)
[0acec5e58ced48c12c28239a20285aa408da5483] Update README and option desc
running  './script.sh'
mise ~/devel/gitlab/triage-ops/undercover/.tool-versions tools: ruby@3.3.3
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Bundle complete! 11 Gemfile dependencies, 35 gems now installed.
Gems in the group 'production' were not installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
Fetching gem metadata from https://rubygems.org/..
Resolving dependencies...
Bundle complete! 4 Gemfile dependencies, 20 gems now installed.
Gems in the group 'production' were not installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
🚨 WARNING: Coverage data is older than your latest changes and results might be incomplete. Re-run tests to update
undercover: 👮‍♂️ some methods have no test coverage! Please add specs for methods listed below
🚨 1) node `block (f)` type: block,
      loc: test.rb:1:1, coverage: 0.0%
1: Dir[File.expand_path("./foo/*.rb", __dir__)].each { |f| require f } hits: n/a
Undercover finished in 0.0041s
Bisecting: 1 revision left to test after this (roughly 1 step)
[036dd0ac6ad1f39cef5217573d5d261c1d2cf8bf] Add file filtering with glob patterns
running  './script.sh'
mise ~/devel/gitlab/triage-ops/undercover/.tool-versions tools: ruby@3.3.3
Bundle complete! 11 Gemfile dependencies, 35 gems now installed.
Gems in the group 'production' were not installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
Bundle complete! 4 Gemfile dependencies, 20 gems now installed.
Gems in the group 'production' were not installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
🚨 WARNING: Coverage data is older than your latest changes and results might be incomplete. Re-run tests to update
undercover: 👮‍♂️ some methods have no test coverage! Please add specs for methods listed below
🚨 1) node `block (f)` type: block,
      loc: test.rb:1:1, coverage: 0.0%
1: Dir[File.expand_path("./foo/*.rb", __dir__)].each { |f| require f } hits: n/a
Undercover finished in 0.0043s
Bisecting: 0 revisions left to test after this (roughly 0 steps)
[e01d298b935b68e58fca3ead3576d360af0f9c05] Report on changed files outside of coverage reports
running  './script.sh'
mise ~/devel/gitlab/triage-ops/undercover/.tool-versions tools: ruby@3.3.3
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
Bundle complete! 11 Gemfile dependencies, 35 gems now installed.
Gems in the group 'production' were not installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
Fetching gem metadata from https://rubygems.org/..
Resolving dependencies...
Bundle complete! 4 Gemfile dependencies, 20 gems now installed.
Gems in the group 'production' were not installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
🚨 WARNING: Coverage data is older than your latest changes and results might be incomplete. Re-run tests to update
undercover: ✅ No coverage is missing in latest changes
Undercover finished in 0.0084s
HEAD is now at e01d298 Report on changed files outside of coverage reports
036dd0ac6ad1f39cef5217573d5d261c1d2cf8bf is the first bad commit
commit 036dd0ac6ad1f39cef5217573d5d261c1d2cf8bf
Author: Jan Grodowski <jgrodowski@gmail.com>
Date:   Tue Dec 3 22:33:02 2024 +0100

    Add file filtering with glob patterns

 lib/undercover.rb            |  15 ++++++++++++++-
 lib/undercover/options.rb    |  30 +++++++++++++++++++++++++-----
 lib/undercover/result.rb     |   7 ++-----
 lib/untested.rb              |   7 +++++++
 spec/cli_spec.rb             |  20 ++++++++++++++++++++
 spec/fixtures/Rakefile       |  15 +++++++++++++++
 spec/fixtures/test.git/index | Bin 494 -> 494 bytes
 spec/report_spec.rb          |  13 +++++++++++--
 undercover.gemspec           |   2 +-
 9 files changed, 95 insertions(+), 14 deletions(-)
 create mode 100644 lib/untested.rb
 create mode 100644 spec/fixtures/Rakefile
```


It's https://github.com/grodowski/undercover/commit/036dd0ac6ad1f39cef5217573d5d261c1d2cf8bf

**Comment by splattael** [comment]
@grodowski I have an ugly looking diff but the spec pass and the reproduction above as well:

```diff
diff --git a/lib/undercover/result.rb b/lib/undercover/result.rb
index fa236ca..87dc98f 100644
--- a/lib/undercover/result.rb
+++ b/lib/undercover/result.rb
@@ -13,7 +13,9 @@ module Undercover
     def initialize(node, file_cov, file_path)
       @node = node
       @coverage = file_cov.select do |ln, _|
-        (node.empty_def? ? ln >= first_line : ln > first_line) && ln < last_line
+        single_line = first_line == last_line && ln == first_line
+        empty_def = node.empty_def? && ln >= first_line
+        (single_line || empty_def) || (ln > first_line && ln < last_line)
       end
       @file_path = file_path
       @flagged = false
```

**Comment by grodowski** [comment]
Thank you so much @splattael for the investigation and proposed fix! My apologies for sneaking this subtle change into an unrelated and badly named commit, this shouldn't happen again. 

Related to https://github.com/grodowski/undercover/commit/036dd0ac6ad1f39cef5217573d5d261c1d2cf8bf, https://github.com/grodowski/imagen_rb/commit/eec7bd4bb7ca502bbb8ee538b2ccf52c7d28cea5 introduces `empty_def?` in the code structure gem and you can check out the unit tests in this commit to see how it behaves.

I'll try to summarise what went wrong: 🐞 for single-line *and* non-empty nodes, the `@coverage` data was blank because the below `select` misses the `first_line == last_line` case 🐞

```ruby
@coverage = file_cov.select do |ln, _|
  (node.empty_def? ? ln >= first_line : ln > first_line) && ln < last_line
end
```

I checked that all existing tests pass with @splattael's patch, would you like to open a pull request? This change could be tested using a fixture setup by extending `spec/result_spec.rb` and then we should be good to release a fixed version 🎉

**Comment by splattael** [comment]
> Thank you so much @splattael for the investigation and proposed fix! My apologies for sneaking this subtle change into an unrelated and badly named commit, this shouldn't happen again.

Don't worry at all :pray: Thanks a ton for this gem :heart: 

> This change could be tested using a fixture setup by extending spec/result_spec.rb and then we should be good to release a fixed version 🎉

I've created https://github.com/grodowski/undercover/pull/218

**Comment by grodowski** [comment]
Just released `v0.6.3` including @splattael's fix. Thanks for contributing!
